File: CS_203/LAB_1/SRC/LinkedList.java
/***************************************************************************
 * <B>Name:</B> Jacob Howarth <BR>
 * CS-203, Spring 2009 <BR>
 * <B>Programming Assignment 1</B> <BR>
 * <B>LinkedList class:</B> This class defines a linked list data structure. <BR>
 * A linked list data structure consists of nodes that hold data and pointers
 * to the next node in the list. These lists are commonly used when memory must
 * be managed dynamically.<BR>
 * <!--Parameters: -->
 * @param T generic data type
 * @author Jacob Howarth 
 * @version 1.5
 ***************************************************************************
 */
public class LinkedList<T> implements List<T> {
    
    /**
     ************************************************************************
     * VARIABLE DECLARATIONS:                                               
     *  Node:
     *  head - A reference to the first element in the list.
     *  tail - A reference to the last element in the list.
     *
     *  Integers:
     *  size - An integer representing the current size of the list.         
     ************************************************************************
     */
    private Node<T> head;
    private Node<T> tail;
    private int size;
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> LinkedList <BR>                                              
     * <B>Description:</B>  A zero argument constructor to create an empty <BR>
     *                      linked list. <BR>   
     * </P>                                                    
     **********************************************************************
     */
    public LinkedList() {
        head = null;
        tail = null;
        size = 0;
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> size <BR>                                              
     * <B>Description:</B>  Returns the current number of Nodes in the <BR>
     *                      linked list. <BR>   
     * </P>
     * <!--Returns: -->
     * @return size - Integer representing the number of nodes in the linked list.                                                    
     **********************************************************************
     */
    public int size() {
        return size;
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> isEmpty <BR>                                            
     * <B>Description:</B> A method that returns a boolean value <BR>      
     *                     representing the following: <BR>
     *                     <DL>                                           
     *                       <DT> <B><I>true</I></B> <DD> - if the linked list
     *                       contains no elements.            
     *                       <DT> <B><I>false</I></B> <DD> - if the linked list 
     *                       has one or more elements.   
     *                     </DL>
     * </P>                                            
     * <!--Returns: -->                                                                                                        
     *  @return boolean - true or false                                             
     **********************************************************************
     */
    public boolean isEmpty() {
        return (head == null);
    }
    
     /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> add <BR>                                            
     * <B>Description:</B> A method that adds a Node holding a data value of <BR>
     *                     generic type, a key to sort the Node by in the list, <BR> 
      *                    and a pointer to the next Node in the list. <BR>
     *                     <BR>
     *                     An IndexOutOfBoundsException is thrown if a Node is <BR>
     *                     added to a negative index or an index greater size. <BR>
     *                     <BR>
     *                     The key string parameter is stored in the Node <BR>
     *                     and is used to relatively sort the Nodes as they <BR>
     *                     are added to the list. <BR>             
     * </P>                                            
     * <!--Parameters: -->
     *  @param element An element of generic data type to be added to the 
     *                 linked list.                                                                                                        
     *  @param index Integer representing the position the element will be
     *               be added to in the linked list. 
     *  @param key A string value used to sort the Nodes in the list.                                             
     **********************************************************************
     */
    public void add(T element, int index) {
        if (index < 0)
            throw new IndexOutOfBoundsException();
        
        Node<T> current = head;
        Node<T> previous = null;
        
         /* This loop goes through the list in reverse order, moving
         * the index while moving the current node pointer. This is done
         * so the whole list doesn't have to be iterated through.
         */
        for (; current != null && index > 0; index--) {
            previous = current; // set the previous node to the current node
                                // before moving current.
            current = current.getNext();
        }
        
        if (index != 0) // if the index isn't zero, this states that an addition
                        // was made outside the current list size.
            throw new IndexOutOfBoundsException();
        
        Node<T> addition = new Node(element);
        addition.setNext(current);
        
        if (previous == null) { // if the list was empty, the head node is
            head = addition; // is the new head node and so is the tail.
            tail = addition;
        }
        else {
            previous.setNext(addition);
            tail = previous.getNext();
        }
        
        size++; // increase size
    }

    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> addLast <BR>
     * <B>Description:</B> Appends elements to the end of the list using  <BR>
     *                     a tail pointer that points the the last element <BR>
     *                     in the list.
     * </P>
     * <!--Parameters: -->
     * @param element An element of a generic data type to be added to the
     *                linked list.
     **********************************************************************
     */
    public void addLast(T element) {

        Node<T> addition = new Node(element); // Create new node
        addition.setNext(null);

        if (head == null) { // If the list is empty, head and tail are the same
            head = addition; // since there is only one element in the list
            tail = addition;
            size++;
        }
        else {
            tail.setNext(addition); // append the new node to the end of the list
            tail = addition;        // and set the new node to be the tail node
            size++;
        }
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> remove <BR>                                            
     * <B>Description:</B> A method that removes a Node from the <BR>
     *                     linked list at the location specified by the index <BR>
     *                     parameter. <BR>
     *                     <BR>
     *                     An IndexOutOfBoundsException is thrown if the index <BR>
     *                     of removal is negative or greater than the current <BR>
     *                     list size. <BR>
     * </P>                                            
     * <!--Parameters: -->                                                                                                        
     *  @param index Integer representing the position of the Node being
     *               removed.                                             
     **********************************************************************
     */
    public void remove(int index) {
        if (index < 0 || index > size)
            throw new IndexOutOfBoundsException();
        
        Node<T> current = head; // start at the head of the list
        Node<T> previous = null;
        
        for (int pos = index; current != null && index > 0; index--) {
            previous = current; // iterate through the list to the position
            current = current.getNext(); // of removal.
        }
        
        if (current == null)
            throw new IndexOutOfBoundsException();
        
        if (previous == null && size == 1) {
            head = previous; // if the list with one node is the node of removal
            tail = previous;
        }
        else if (previous == null && size > 1) {
            current = current.getNext(); // use this code if the index of
            head = current;              // removal is at the head to prevent
                                        // deletion of an entire list.
        }
        else if (current.getNext() == null) {
            tail = previous;
            previous.setNext(current.getNext());
        }
        else {
            previous.setNext(current.getNext());
        }
        size--; // decrease the size
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> get <BR>                                            
     * <B>Description:</B> A method that returns the first instance of an <BR>
     *                     element in the linked list. The iterating is done <BR>
     *                     in reverse order for efficiency. <BR>
     *                     <BR>
     *                     An IndexOutOfBoundsException is thrown if the index <BR> 
     *                     is negative or greater than the size and if the <BR>
     *                     the current node pointer is null. <BR>
     * </P>                                            
     * <!--Parameters: -->                                                                                                        
     *  @param index Integer representing the position of the element.
     * <!--Returns: -->
     *  @return T - element of generic type.                                               
     **********************************************************************
     */
    public T get(int index) {
        Node<T> current = head; // start iterating at the head of the linked list.
        
        for(; current != null && index > 0; index--)
            current = current.getNext(); // iterate through the linked list
                                        // stopping at the index of the element
                                        // being retrieved.
        if (current == null || index != 0)
            throw new IndexOutOfBoundsException();
        
        return current.getElement(); // return the element at the specified index. 
    }

    public T getLast() {
        return tail.getElement();
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> indexOf <BR>                                            
     * <B>Description:</B> A method that returns the index of the first <BR> 
     *                     instance of an element in the linked list. <BR>
     *                     If the element is <I>not found</I> the method <BR>
     *                     returns -1. <BR>
     * </P>                                            
     * <!--Parameters: -->                                                                                                        
     *  @param element element of generic type being searched for.
     * <!--Returns: -->
     *  @return index - Integer representing the index of the element being
     *                  searched for if found. Returns -1 if not found.                                               
     **********************************************************************
     */
    public int indexOf(T element) {
        final int NOT_FOUND = -1; // return if the element is not in the list.
        int count = 0; // a counter to return the index of the element being 
                        // searched for.
        LinkedListIterator<T> iterator = iterator(); // create an iterator to
        T nodeElement;                               // iterate through each node.
        
        while (iterator.hasNext()) {
            nodeElement = iterator.next(); // grad the data element from the Node.
            if (nodeElement.equals(element)) // test if the elements are equal.
                return count;
            count++;
        }
        
        return NOT_FOUND;
    }
    
     /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> iterator <BR>                                            
     * <B>Description:</B> A method that returns a linked list iterator instance <BR>
     *                     which can be used to iterate through each node <BR>
     *                     currently in the linked list. <BR>
     * </P>                                            
     * <!--Returns: -->
     *  @return LinkedListIterator&lt;T&gt; - iterator that iterates through 
     *                                   the nodes in the list.                                               
     **********************************************************************
     */
    public LinkedListIterator<T> iterator() {
        return (new LinkedListIterator<T>(size, head));
    }
    
    /**
     **********************************************************************
     * <P align="left">
     * <B>Method:</B> toString <BR>                                                   
     * <B>Description:</B> A method that prints the data in the linked list <BR>
     * </P>                                                                              
     * <!--Returns: -->                                                    
     *  @return result - A string representing the data in the linked list.                                                     
     **********************************************************************
     */ 
    @Override
    public String toString() {
        LinkedListIterator<T> iterator = iterator();
        String result = "";
        
        while (iterator.hasNext())
            result += iterator.next().toString() + "\n";
        
        return result;
    }
    
}

File: CS_202/LAB_1/LinkedList.c
/* File:   LinkedList.c
 * Author: Jacob S. Howarth
 * Class: CS-202: Systems Programming Concepts
 *
 * Description: This program allows a user to create
 * a linked list of integers.
 *
 * Created on April 8, 2009, 7:09 PM
 */

// HEADER FILE DECLARATIONS
#include <stdio.h>
#include <stdlib.h>

/* GLOBAL VARIABLES
 * struct intNode: A structure representing a linked list node
 *                 that contains the following items:
 *
 *                 int value = the integer to be stored in the node.
 *                 struct intNode* next = A pointer/reference to the
 *                 next node in the list.
 *
 * struct* head = A reference to the first node of the list.
 *                This node will only contain the reference
 *                address to the first node in the list.
 * struct* tail = A reference to the last node in the list.
 *                This node will only contain a reference to the
 *                last node of the list.
 */
struct intNode {
    int value;
    struct intNode* next;
}*head = NULL, *tail = NULL;

/* FUNCTION PROTOTYPE DECLARATIONS
 * void add(int element); // adds a new node to the end of the list
 * void printList(int* sizeOf); // prints each node in the list.
 */
void add(int element);
void printList(int* sizeOf);

int main(int argc, char** argv) {
    int sizeOf, value;

    // Get the size of the linked list from the user.
    printf("Enter size of list: ");
    scanf("%i", &sizeOf);
    
    // Ask the user to enter each integer for the linked list.
    int i;
    for (i = 0; i < sizeOf; i++) {
    printf("Enter an integer to store: ");
        scanf("%i", &value);
        add(value); // take the entered integers and create new nodes for each
                    // Do this until the list size, specified by the user is
                    // reached.
    }

    printList(&sizeOf); // print the list

    return (EXIT_SUCCESS);
}

/* Function: add
 * Parameters: int element - integer to be added to the list
 *
 * Description: Creates a new node with value of the parameter and
 *              adds it to the end of the list. If the list is empty
 *              or null, the new node is referenced by the head and
 *              tail pointers.
 *
 * Returns: void
 */
void add(int element) {

    // Allocate a space in the heap for a new node and return
    // a pointer to that node. Dereference the pointer to store
    // the parameter and set it's next pointer to null.
    struct intNode* newNode = malloc(sizeof(struct intNode));
    newNode->value = element;
    newNode->next = NULL;

    // If the list is empty and the reference of the node returned
    // by malloc is not null, set the head and tail pointers to
    // reference the new node.
    if (head == NULL && newNode != NULL) {
        head = newNode;
        tail = newNode;
    }
    else {
        // otherwise, set the next pointer of the end node to point
        // to the new node and since the tail references the end node
        // assign the reference of the end node to the tail pointer.
        tail->next = newNode;
        tail = newNode;
    }

}

/* Function: printList
 * Parameters: int* sizeOf - integer pointer that contains the reference
 *                           address of sizeOf to avoid deep copying.
 *
 * Description: Prints the list in the following format:
 *
 *              1) The first node is printed with an indicator of "Head"
 *                 plus the first nodes value.
 *              2) The following nodes up to size of the list - 1 will
 *      `          be printed with the indicator "Node" plus it's value.
 *              3) The last node is printed with an indicator of "Tail"
 *                 plus its value.
 *
 * Returns: void
 */
void printList(int* sizeOf) {
    // Create a new pointers to point to the beginning of
    // the list and allow list iteration.
    struct intNode* current = head;

    printf("\nHere is the list:\n");

    int i;
    for (i = 0; current != NULL; i++) {
        if (i == 0) // first node is printed with "Head" indicator plus its values.
            fprintf("Head  ->%3i\n", current->value);
        else if (*sizeOf - 1 == i) // last node is printed with "Tail" indicator
                                    // plus its value.
            fprintf("Tail  ->%3i\n", current->value);
        else
            // subsequent nodes are printed with "Node"
            // indicator plus their values.
            fprintf("Node%2i->%3i\n", i, current->value);

        current = current->next; // advance to the next node in the list.
    }
}

